README.md
# Terraform Provider: Kerberos

Terraform Provider for Kerberos resources.

## Requirements

- [Terraform](https://www.terraform.io/downloads.html) >= 1.0
- [Go](https://golang.org/doc/install) >= 1.18

## Building The Provider

1. Clone the repository
1. Enter the repository directory
1. Build the provider using the Go `install` command:

```shell
go install
```

## Adding Dependencies

This provider uses [Go modules](https://github.com/golang/go/wiki/Modules).
Please see the Go documentation for the most up to date information about using Go modules.

To add a new dependency `github.com/author/dependency` to your Terraform provider:

```shell
go get github.com/author/dependency
go mod tidy
```

Then commit the changes to `go.mod` and `go.sum`.

## Using the provider

Install the provider by adding it to the `required_provider` block of your `terraform` configuration:

```hcl
terraform {
  required_providers {
    kerberos = {
      source = "zanecodes/kerberos"
      version = "~> 0.1"
    }
  }
}
```

## Developing the Provider

If you wish to work on the provider, you'll first need [Go](http://www.golang.org) installed on your machine (see [Requirements](#requirements) above).

To compile the provider, run `go install`. This will build the provider and put the provider binary in the `$GOPATH/bin` directory.

To generate or update documentation, run `go generate`.

In order to run the full suite of Acceptance tests, run `make testacc`.

```shell
make testacc
```


CHANGELOG.md
## 0.2.0

BREAKING CHANGES:

* Provider has been renamed to `kerberos`
* resource/keytab\_file: Renamed to `kerberos_keytab`

FEATURES:

* New Data Source: `kerberos_token`

## 0.1.0

FEATURES:

* New Resource: `keytab_file`


internal/provider/provider_test.go
package provider

import (
	"testing"

	"github.com/hashicorp/terraform-plugin-framework/providerserver"
	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
)

// testAccProtoV6ProviderFactories are used to instantiate a provider during
// acceptance testing. The factory function will be invoked for every Terraform
// CLI command executed to create a provider server to which the CLI can
// reattach.
var testAccProtoV6ProviderFactories = map[string]func() (tfprotov6.ProviderServer, error){
	"kerberos": providerserver.NewProtocol6WithError(New("test")()),
}

func testAccPreCheck(t *testing.T) {
	// You can add code here to run prior to any test case execution, for example assertions
	// about the appropriate environment variables being set are common to see in a pre-check
	// function.
}


internal/provider/file_resource_test.go
package provider

import (
	"encoding/base64"
	"fmt"
	"testing"
	"time"

	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
	"github.com/jcmturner/gokrb5/v8/iana/etypeID"
	"github.com/jcmturner/gokrb5/v8/keytab"
)

func TestAccFileResource(t *testing.T) {
	first_keytab := keytab.New()

	second_keytab := keytab.New()
	if err := second_keytab.AddEntry("principal", "realm.com", "key", time.Unix(0, 0), 0, etypeID.RC4_HMAC); err != nil {
		t.Fatal(err.Error())
		return
	}

	third_keytab := keytab.New()
	if err := third_keytab.AddEntry("principal", "realm.com", "key", time.Unix(0, 0), 0, etypeID.RC4_HMAC); err != nil {
		t.Fatal(err.Error())
		return
	}
	if err := third_keytab.AddEntry("principal two", "realm-two.com", "key two", time.Unix(1, 0), 1, etypeID.AES128_CTS_HMAC_SHA1_96); err != nil {
		t.Fatal(err.Error())
		return
	}

	resource.Test(t, resource.TestCase{
		PreCheck:                 func() { testAccPreCheck(t) },
		ProtoV6ProviderFactories: testAccProtoV6ProviderFactories,
		Steps: []resource.TestStep{
			// Create and Read testing
			{
				Config: `
resource "kerberos_file" "test" {
  entry {
    principal = "principal"
    realm = "realm.com"
    key = "key"
    key_version = 0
    encryption_type = "rc4-hmac"
  }
}
`,
				PreConfig: func() {
					if err := first_keytab.AddEntry("principal", "realm.com", "key", time.Now().Truncate(time.Second), 0, etypeID.RC4_HMAC); err != nil {
						t.Fatal(err.Error())
						return
					}
				},
				Check: resource.ComposeAggregateTestCheckFunc(
					resource.TestCheckResourceAttrSet("kerberos_file.test", "id"),
					resource.TestCheckResourceAttrWith("kerberos_file.test", "content_base64", testAccCheckKeytabContent(t, first_keytab)),
				),
			},
			// Update and Read testing
			{
				Config: `
resource "kerberos_file" "test" {
  entry {
    principal = "principal"
    realm = "realm.com"
    key = "key"
    key_version = 0
    encryption_type = "rc4-hmac"
    timestamp = "1970-01-01T00:00:00Z"
  }
}
`,
				Check: resource.ComposeAggregateTestCheckFunc(
					resource.TestCheckResourceAttrWith("kerberos_file.test", "content_base64", testAccCheckKeytabContent(t, second_keytab)),
				),
			},
			{
				Config: `
resource "kerberos_file" "test" {
  entry {
    principal = "principal"
    realm = "realm.com"
    key = "key"
    key_version = 0
    encryption_type = "rc4-hmac"
    timestamp = "1970-01-01T00:00:00Z"
  }
  entry {
    principal = "principal two"
    realm = "realm-two.com"
    key = "key two"
    key_version = 1
    encryption_type = "aes128-sha1"
    timestamp = "1970-01-01T00:00:01Z"
  }
}
`,
				Check: resource.ComposeAggregateTestCheckFunc(
					resource.TestCheckResourceAttrWith("kerberos_file.test", "content_base64", testAccCheckKeytabContent(t, third_keytab)),
				),
			},
			// Delete testing automatically occurs in TestCase
		},
	})
}

func testAccCheckKeytabContent(_ *testing.T, expected *keytab.Keytab) resource.CheckResourceAttrWithFunc {
	return func(actual_value string) error {
		expected_bytes, err := expected.Marshal()

		if err != nil {
			return err
		}

		expected_value := base64.StdEncoding.EncodeToString(expected_bytes)

		if actual_value != expected_value {
			actual_bytes, err := base64.StdEncoding.DecodeString(actual_value)

			if err != nil {
				return err
			}

			actual := keytab.New()
			err = actual.Unmarshal(actual_bytes)

			if err != nil {
				return err
			}

			// Allow for up to 5 seconds of slop in timestamps during testing
			for i, entry := range expected.Entries {
				if actual.Entries[i].Timestamp.Sub(entry.Timestamp).Seconds() < 5.0 {
					actual.Entries[i].Timestamp = entry.Timestamp
				}
			}

			actual_bytes, err = actual.Marshal()

			if err != nil {
				return err
			}

			actual_value = base64.StdEncoding.EncodeToString(actual_bytes)

			if actual_value == expected_value {
				return nil
			}

			expected_json, err := expected.JSON()

			if err != nil {
				return err
			}

			actual_json, err := actual.JSON()

			if err != nil {
				return err
			}

			return fmt.Errorf("Expected keytab:\n%s\nActual keytab:\n%s", expected_json, actual_json)
		}

		return nil
	}
}


internal/provider/keytab_resource_test.go
package provider

import (
	"encoding/base64"
	"fmt"
	"testing"
	"time"

	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
	"github.com/jcmturner/gokrb5/v8/iana/etypeID"
	"github.com/jcmturner/gokrb5/v8/keytab"
)

func TestAccKeytabResource(t *testing.T) {
	first_keytab := keytab.New()

	second_keytab := keytab.New()
	if err := second_keytab.AddEntry("principal", "realm.com", "key", time.Unix(0, 0), 0, etypeID.RC4_HMAC); err != nil {
		t.Fatal(err.Error())
		return
	}

	third_keytab := keytab.New()
	if err := third_keytab.AddEntry("principal", "realm.com", "key", time.Unix(0, 0), 0, etypeID.RC4_HMAC); err != nil {
		t.Fatal(err.Error())
		return
	}
	if err := third_keytab.AddEntry("principal two", "realm-two.com", "key two", time.Unix(1, 0), 1, etypeID.AES128_CTS_HMAC_SHA1_96); err != nil {
		t.Fatal(err.Error())
		return
	}

	resource.Test(t, resource.TestCase{
		PreCheck:                 func() { testAccPreCheck(t) },
		ProtoV6ProviderFactories: testAccProtoV6ProviderFactories,
		Steps: []resource.TestStep{
			// Create and Read testing
			{
				Config: `
resource "kerberos_keytab" "test" {
  entry {
    principal = "principal"
    realm = "realm.com"
    key = "key"
    key_version = 0
    encryption_type = "rc4-hmac"
  }
}
`,
				PreConfig: func() {
					if err := first_keytab.AddEntry("principal", "realm.com", "key", time.Now().Truncate(time.Second), 0, etypeID.RC4_HMAC); err != nil {
						t.Fatal(err.Error())
						return
					}
				},
				Check: resource.ComposeAggregateTestCheckFunc(
					resource.TestCheckResourceAttrSet("kerberos_keytab.test", "id"),
					resource.TestCheckResourceAttrWith("kerberos_keytab.test", "content_base64", testAccCheckKeytabContent(t, first_keytab)),
				),
			},
			// Update and Read testing
			{
				Config: `
resource "kerberos_keytab" "test" {
  entry {
    principal = "principal"
    realm = "realm.com"
    key = "key"
    key_version = 0
    encryption_type = "rc4-hmac"
    timestamp = "1970-01-01T00:00:00Z"
  }
}
`,
				Check: resource.ComposeAggregateTestCheckFunc(
					resource.TestCheckResourceAttrWith("kerberos_keytab.test", "content_base64", testAccCheckKeytabContent(t, second_keytab)),
				),
			},
			{
				Config: `
resource "kerberos_keytab" "test" {
  entry {
    principal = "principal"
    realm = "realm.com"
    key = "key"
    key_version = 0
    encryption_type = "rc4-hmac"
    timestamp = "1970-01-01T00:00:00Z"
  }
  entry {
    principal = "principal two"
    realm = "realm-two.com"
    key = "key two"
    key_version = 1
    encryption_type = "aes128-sha1"
    timestamp = "1970-01-01T00:00:01Z"
  }
}
`,
				Check: resource.ComposeAggregateTestCheckFunc(
					resource.TestCheckResourceAttrWith("kerberos_keytab.test", "content_base64", testAccCheckKeytabContent(t, third_keytab)),
				),
			},
			// Delete testing automatically occurs in TestCase
		},
	})
}

func testAccCheckKeytabContent(_ *testing.T, expected *keytab.Keytab) resource.CheckResourceAttrWithFunc {
	return func(actual_value string) error {
		expected_bytes, err := expected.Marshal()

		if err != nil {
			return err
		}

		expected_value := base64.StdEncoding.EncodeToString(expected_bytes)

		if actual_value != expected_value {
			actual_bytes, err := base64.StdEncoding.DecodeString(actual_value)

			if err != nil {
				return err
			}

			actual := keytab.New()
			err = actual.Unmarshal(actual_bytes)

			if err != nil {
				return err
			}

			// Allow for up to 5 seconds of slop in timestamps during testing
			for i, entry := range expected.Entries {
				if actual.Entries[i].Timestamp.Sub(entry.Timestamp).Seconds() < 5.0 {
					actual.Entries[i].Timestamp = entry.Timestamp
				}
			}

			actual_bytes, err = actual.Marshal()

			if err != nil {
				return err
			}

			actual_value = base64.StdEncoding.EncodeToString(actual_bytes)

			if actual_value == expected_value {
				return nil
			}

			expected_json, err := expected.JSON()

			if err != nil {
				return err
			}

			actual_json, err := actual.JSON()

			if err != nil {
				return err
			}

			return fmt.Errorf("Expected keytab:\n%s\nActual keytab:\n%s", expected_json, actual_json)
		}

		return nil
	}
}


main.go
package main

import (
	"context"
	"flag"
	"log"

	"github.com/hashicorp/terraform-plugin-framework/providerserver"
	"github.com/zanecodes/terraform-provider-kerberos/internal/provider"
)

// Run "go generate" to format example terraform files and generate the docs for the registry/website

// If you do not have terraform installed, you can remove the formatting command, but its suggested to
// ensure the documentation is formatted properly.
//go:generate terraform fmt -recursive ./examples/

// Run the docs generation tool, check its repository for more information on how it works and how docs
// can be customized.
//go:generate go run github.com/hashicorp/terraform-plugin-docs/cmd/tfplugindocs

var (
	// these will be set by the goreleaser configuration
	// to appropriate values for the compiled binary.
	version string = "dev"

	// goreleaser can pass other information to the main package, such as the specific commit
	// https://goreleaser.com/cookbooks/using-main.version/
)

func main() {
	var debug bool

	flag.BoolVar(&debug, "debug", false, "set to true to run the provider with support for debuggers like delve")
	flag.Parse()

	opts := providerserver.ServeOpts{
		Address: "registry.terraform.io/zanecodes/kerberos",
		Debug:   debug,
	}

	err := providerserver.Serve(context.Background(), provider.New(version), opts)

	if err != nil {
		log.Fatal(err.Error())
	}
}


internal/provider/provider.go
package provider

import (
	"context"

	"github.com/hashicorp/terraform-plugin-framework/datasource"
	"github.com/hashicorp/terraform-plugin-framework/provider"
	"github.com/hashicorp/terraform-plugin-framework/resource"
)

// Ensure KerberosProvider satisfies various provider interfaces.
var _ provider.Provider = &KerberosProvider{}

// KerberosProvider defines the provider implementation.
type KerberosProvider struct {
	// version is set to the provider version on release, "dev" when the
	// provider is built and ran locally, and "test" when running acceptance
	// testing.
	version string
}

// KerberosProviderModel describes the provider data model.
type KerberosProviderModel struct{}

func (p *KerberosProvider) Metadata(ctx context.Context, req provider.MetadataRequest, resp *provider.MetadataResponse) {
	resp.TypeName = "kerberos"
	resp.Version = p.version
}

func (p *KerberosProvider) Schema(ctx context.Context, req provider.SchemaRequest, resp *provider.SchemaResponse) {
}

func (p *KerberosProvider) Configure(ctx context.Context, req provider.ConfigureRequest, resp *provider.ConfigureResponse) {
}

func (p *KerberosProvider) Resources(ctx context.Context) []func() resource.Resource {
	return []func() resource.Resource{
		NewKeytabResource,
	}
}

func (p *KerberosProvider) DataSources(ctx context.Context) []func() datasource.DataSource {
	return []func() datasource.DataSource{
		NewTokenDataSource,
	}
}

func New(version string) func() provider.Provider {
	return func() provider.Provider {
		return &KerberosProvider{
			version: version,
		}
	}
}


internal/provider/file_resource.go
package provider

import (
	"context"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"math"
	"time"

	"github.com/hashicorp/terraform-plugin-framework-validators/int64validator"
	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
	"github.com/hashicorp/terraform-plugin-framework/resource"
	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
	"github.com/hashicorp/terraform-plugin-framework/resource/schema/objectplanmodifier"
	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
	"github.com/hashicorp/terraform-plugin-framework/types"
	"github.com/hashicorp/terraform-plugin-log/tflog"
	"github.com/jcmturner/gokrb5/v8/iana/etypeID"
	"github.com/jcmturner/gokrb5/v8/keytab"

	"github.com/zanecodes/terraform-provider-kerberos/internal/validators/timevalidator"
)

// Ensure provider defined types fully satisfy framework interfaces.
var _ resource.Resource = &FileResource{}

func NewFileResource() resource.Resource {
	return &FileResource{}
}

// FileResource defines the resource implementation.
type FileResource struct {
}

// FileResourceModel describes the resource data model.
type FileResourceModel struct {
	Entries       []FileEntryModel `tfsdk:"entry"`
	ContentBase64 types.String     `tfsdk:"content_base64"`
	Id            types.String     `tfsdk:"id"`
}

type FileEntryModel struct {
	Principal      types.String `tfsdk:"principal"`
	Realm          types.String `tfsdk:"realm"`
	Key            types.String `tfsdk:"key"`
	KeyVersion     types.Int64  `tfsdk:"key_version"`
	EncryptionType types.String `tfsdk:"encryption_type"`
	Timestamp      types.String `tfsdk:"timestamp"`
}

func (r *FileResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
	resp.TypeName = req.ProviderTypeName + "_file"
}

func (r *FileResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
	supported_etypes := make([]string, 0)

	for k := range etypeID.ETypesByName {
		if etypeID.EtypeSupported(k) != 0 {
			supported_etypes = append(supported_etypes, k)
		}
	}

	resp.Schema = schema.Schema{
		// This description is used by the documentation generator and the language server.
		MarkdownDescription: "Manages a Kerberos keytab file stored in the Terraform state. The generated keytab contents can be read from the `content_base64` attribute.",

		Blocks: map[string]schema.Block{
			"entry": schema.ListNestedBlock{
				NestedObject: schema.NestedBlockObject{
					Attributes: map[string]schema.Attribute{
						"principal": schema.StringAttribute{
							MarkdownDescription: "The name of the Kerberos principal to which the key belongs, not including the realm.",
							Required:            true,
						},
						"realm": schema.StringAttribute{
							MarkdownDescription: "The realm to which the Kerberos principal belongs.",
							Required:            true,
						},
						"key": schema.StringAttribute{
							MarkdownDescription: "The key belonging to the Kerberos principal.",
							Required:            true,
							Sensitive:           true,
						},
						"key_version": schema.Int64Attribute{
							MarkdownDescription: "The version number of the key.",
							Required:            true,
							Validators: []validator.Int64{
								int64validator.Between(0, math.MaxUint8),
							},
						},
						"encryption_type": schema.StringAttribute{
							MarkdownDescription: "The encryption type to use for the key. Must be one of: `aes128-cts-hmac-sha1-96`/`aes128-cts`/`aes128-sha1`, `aes256-cts-hmac-sha1-96`/`aes256-cts`/`aes256-sha1`, `aes128-cts-hmac-sha256-128`/`aes128-sha2`, `aes256-cts-hmac-sha384-192`/`aes256-sha2`, `des3-cbc-sha1-kd`, or `arcfour-hmac`/`rc4-hmac`/`arcfour-hmac-md5`.",
							Required:            true,
							Validators: []validator.String{
								stringvalidator.OneOf(supported_etypes...),
							},
						},
						"timestamp": schema.StringAttribute{
							MarkdownDescription: "The creation timestamp for the Keytab entry in [RFC3339](https://datatracker.ietf.org/doc/html/rfc3339#section-5.8) format. Defaults to the current time, but may be set to a specific time for better reproducibility.",
							Optional:            true,
							Computed:            true,
							Validators: []validator.String{
								timevalidator.IsRFC3339Time(),
							},
						},
					},

					PlanModifiers: []planmodifier.Object{
						objectplanmodifier.RequiresReplaceIfConfigured(),
					},
				},
			},
		},

		Attributes: map[string]schema.Attribute{
			"content_base64": schema.StringAttribute{
				MarkdownDescription: "The base64 encoded keytab contents.",
				Computed:            true,
				Sensitive:           true,
			},
			"id": schema.StringAttribute{
				MarkdownDescription: "The SHA256 hash of the binary keytab contents.",
				Computed:            true,
			},
		},
	}
}

func (r *FileResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
}

func (r *FileResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
	var data *FileResourceModel

	now := time.Now()

	// Read Terraform plan data into the model
	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)

	if resp.Diagnostics.HasError() {
		return
	}

	keytab := keytab.New()

	for i, entry := range data.Entries {
		var timestamp time.Time

		if entry.Timestamp.IsUnknown() {
			timestamp = now
			data.Entries[i].Timestamp = types.StringValue(timestamp.Format(time.RFC3339))
		} else {
			var err error
			timestamp, err = time.Parse(time.RFC3339, entry.Timestamp.ValueString())

			if err != nil {
				resp.Diagnostics.AddError("Invalid timestamp", err.Error())
				return
			}
		}

		if err := keytab.AddEntry(entry.Principal.ValueString(), entry.Realm.ValueString(), entry.Key.ValueString(), timestamp, uint8(entry.KeyVersion.ValueInt64()), etypeID.EtypeSupported(entry.EncryptionType.ValueString())); err != nil {
			resp.Diagnostics.AddError("Invalid keytab entry", err.Error())
			return
		}
	}

	bytes, err := keytab.Marshal()

	if err != nil {
		resp.Diagnostics.AddError("Unable to generate keytab", err.Error())
		return
	}

	data.ContentBase64 = types.StringValue(base64.StdEncoding.EncodeToString(bytes))

	sum := sha256.Sum256(bytes)
	data.Id = types.StringValue(fmt.Sprintf("%x", sum[:]))

	// Write logs using the tflog package
	// Documentation: https://terraform.io/plugin/log
	tflog.Trace(ctx, "created a resource")

	// Save data into Terraform state
	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
}

func (r *FileResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
	var data *FileResourceModel

	// Read Terraform prior state data into the model
	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)

	if resp.Diagnostics.HasError() {
		return
	}

	// Save updated data into Terraform state
	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
}

func (r *FileResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
	var data *FileResourceModel

	// Read Terraform plan data into the model
	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)

	if resp.Diagnostics.HasError() {
		return
	}

	// Save updated data into Terraform state
	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
}

func (r *FileResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
	var data *FileResourceModel

	// Read Terraform prior state data into the model
	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)

	if resp.Diagnostics.HasError() {
		return
	}
}


internal/provider/keytab_resource.go
package provider

import (
	"context"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"math"
	"time"

	"github.com/hashicorp/terraform-plugin-framework-validators/int64validator"
	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
	"github.com/hashicorp/terraform-plugin-framework/resource"
	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
	"github.com/hashicorp/terraform-plugin-framework/resource/schema/objectplanmodifier"
	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
	"github.com/hashicorp/terraform-plugin-framework/types"
	"github.com/hashicorp/terraform-plugin-log/tflog"
	"github.com/jcmturner/gokrb5/v8/iana/etypeID"
	"github.com/jcmturner/gokrb5/v8/keytab"

	"github.com/zanecodes/terraform-provider-kerberos/internal/validators/timevalidator"
)

// Ensure provider defined types fully satisfy framework interfaces.
var _ resource.Resource = &KeytabResource{}

func NewKeytabResource() resource.Resource {
	return &KeytabResource{}
}

// KeytabResource defines the resource implementation.
type KeytabResource struct {
}

// KeytabResourceModel describes the resource data model.
type KeytabResourceModel struct {
	Entries       []KeytabEntryModel `tfsdk:"entry"`
	ContentBase64 types.String       `tfsdk:"content_base64"`
	Id            types.String       `tfsdk:"id"`
}

type KeytabEntryModel struct {
	Principal      types.String `tfsdk:"principal"`
	Realm          types.String `tfsdk:"realm"`
	Key            types.String `tfsdk:"key"`
	KeyVersion     types.Int64  `tfsdk:"key_version"`
	EncryptionType types.String `tfsdk:"encryption_type"`
	Timestamp      types.String `tfsdk:"timestamp"`
}

func (r *KeytabResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
	resp.TypeName = req.ProviderTypeName + "_keytab"
}

func (r *KeytabResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
	supported_etypes := make([]string, 0)

	for k := range etypeID.ETypesByName {
		if etypeID.EtypeSupported(k) != 0 {
			supported_etypes = append(supported_etypes, k)
		}
	}

	resp.Schema = schema.Schema{
		// This description is used by the documentation generator and the language server.
		MarkdownDescription: "Manages a Kerberos keytab file stored in the Terraform state. The generated keytab contents can be read from the `content_base64` attribute.",

		Blocks: map[string]schema.Block{
			"entry": schema.ListNestedBlock{
				NestedObject: schema.NestedBlockObject{
					Attributes: map[string]schema.Attribute{
						"principal": schema.StringAttribute{
							MarkdownDescription: "The name of the Kerberos principal to which the key belongs, not including the realm.",
							Required:            true,
						},
						"realm": schema.StringAttribute{
							MarkdownDescription: "The realm to which the Kerberos principal belongs.",
							Required:            true,
						},
						"key": schema.StringAttribute{
							MarkdownDescription: "The key belonging to the Kerberos principal.",
							Required:            true,
							Sensitive:           true,
						},
						"key_version": schema.Int64Attribute{
							MarkdownDescription: "The version number of the key.",
							Required:            true,
							Validators: []validator.Int64{
								int64validator.Between(0, math.MaxUint8),
							},
						},
						"encryption_type": schema.StringAttribute{
							MarkdownDescription: "The encryption type to use for the key. Must be one of: `aes128-cts-hmac-sha1-96`/`aes128-cts`/`aes128-sha1`, `aes256-cts-hmac-sha1-96`/`aes256-cts`/`aes256-sha1`, `aes128-cts-hmac-sha256-128`/`aes128-sha2`, `aes256-cts-hmac-sha384-192`/`aes256-sha2`, `des3-cbc-sha1-kd`, or `arcfour-hmac`/`rc4-hmac`/`arcfour-hmac-md5`.",
							Required:            true,
							Validators: []validator.String{
								stringvalidator.OneOf(supported_etypes...),
							},
						},
						"timestamp": schema.StringAttribute{
							MarkdownDescription: "The creation timestamp for the Keytab entry in [RFC3339](https://datatracker.ietf.org/doc/html/rfc3339#section-5.8) format. Defaults to the current time, but may be set to a specific time for better reproducibility.",
							Optional:            true,
							Computed:            true,
							Validators: []validator.String{
								timevalidator.IsRFC3339Time(),
							},
						},
					},

					PlanModifiers: []planmodifier.Object{
						objectplanmodifier.RequiresReplaceIfConfigured(),
					},
				},
			},
		},

		Attributes: map[string]schema.Attribute{
			"content_base64": schema.StringAttribute{
				MarkdownDescription: "The base64 encoded keytab contents.",
				Computed:            true,
				Sensitive:           true,
			},
			"id": schema.StringAttribute{
				MarkdownDescription: "The SHA256 hash of the binary keytab contents.",
				Computed:            true,
			},
		},
	}
}

func (r *KeytabResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
}

func (r *KeytabResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
	var data *KeytabResourceModel

	now := time.Now()

	// Read Terraform plan data into the model
	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)

	if resp.Diagnostics.HasError() {
		return
	}

	keytab := keytab.New()

	for i, entry := range data.Entries {
		var timestamp time.Time

		if entry.Timestamp.IsUnknown() {
			timestamp = now
			data.Entries[i].Timestamp = types.StringValue(timestamp.Format(time.RFC3339))
		} else {
			var err error
			timestamp, err = time.Parse(time.RFC3339, entry.Timestamp.ValueString())

			if err != nil {
				resp.Diagnostics.AddError("Invalid timestamp", err.Error())
				return
			}
		}

		if err := keytab.AddEntry(entry.Principal.ValueString(), entry.Realm.ValueString(), entry.Key.ValueString(), timestamp, uint8(entry.KeyVersion.ValueInt64()), etypeID.EtypeSupported(entry.EncryptionType.ValueString())); err != nil {
			resp.Diagnostics.AddError("Invalid keytab entry", err.Error())
			return
		}
	}

	bytes, err := keytab.Marshal()

	if err != nil {
		resp.Diagnostics.AddError("Unable to generate keytab", err.Error())
		return
	}

	data.ContentBase64 = types.StringValue(base64.StdEncoding.EncodeToString(bytes))

	sum := sha256.Sum256(bytes)
	data.Id = types.StringValue(fmt.Sprintf("%x", sum[:]))

	// Write logs using the tflog package
	// Documentation: https://terraform.io/plugin/log
	tflog.Trace(ctx, "created a resource")

	// Save data into Terraform state
	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
}

func (r *KeytabResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
	var data *KeytabResourceModel

	// Read Terraform prior state data into the model
	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)

	if resp.Diagnostics.HasError() {
		return
	}

	// Save updated data into Terraform state
	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
}

func (r *KeytabResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
	var data *KeytabResourceModel

	// Read Terraform plan data into the model
	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)

	if resp.Diagnostics.HasError() {
		return
	}

	// Save updated data into Terraform state
	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
}

func (r *KeytabResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
	var data *KeytabResourceModel

	// Read Terraform prior state data into the model
	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)

	if resp.Diagnostics.HasError() {
		return
	}
}
